### PR TITLE
Log custom creation script execution in resource logger for `WithCreationScript(...)`

### LINKED ISSUE #9228: `WithCreationScript(...)` should provide feedback in resource logger.
Image

At the moment if I have a problem with my database creation script I don't get any feedback. We should show what block is being executed and what the result was.

### PR DESCRIPTION
**Pull request by mitchdenny**
Fixes #9228

## Summary

When a custom database creation script is provided via `WithCreationScript(...)`, the resource logger now emits informational messages showing what is being executed and what the outcome was. Previously only generic "Creating database" / "created successfully" / "Failed to create database" messages were logged, with no visibility into the custom script itself.

### SQL Server (`Aspire.Hosting.SqlServer`)
- Logs that custom script execution is starting
- For each `GO`-delimited batch: logs the batch content and rows affected per execution
- Logs overall script completion

### PostgreSQL (`Aspire.Hosting.PostgreSQL`)
- Logs custom script content before execution
- Logs rows affected on completion
- Default (no `WithCreationScript`) path is unchanged

### MySQL (`Aspire.Hosting.MySql`)
- Logs custom script content before execution
- Logs rows affected on completion
- Default (no `WithCreationScript`) path is unchanged

### Azure Kusto (`Aspire.Hosting.Azure.Kusto`)
- Detects custom annotation vs. default generated command
- When a custom script is present: logs script content before execution and logs completion

## Testing

All non-Docker unit tests pass:
- SQL Server: 65/65 ✅
- PostgreSQL: 106/106 ✅
- MySQL: 65/65 ✅
- Kusto (unit tests): 52/52 ✅

### CODE REVIEW

**Comment by mitchdenny** [comment]
## Local Test Results

Ran the non-Docker unit/API tests for all four affected integrations after the change:

| Test Project | Result | Count |
|---|---|---|
|  | ✅ Passed | 65/65 |
|  | ✅ Passed | 106/106 |
|  | ✅ Passed | 65/65 |
|  (unit only, excl. ) | ✅ Passed | 52/52 |

The 4  failures require a running Kusto Docker emulator and are pre-existing infrastructure issues unrelated to this change.

No test files were modified — only the four source files.

File: src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs [hunk 1]
```diff
@@ -291,7 +292,18 @@ private static async Task CreateDatabaseAsync(ICslAdminProvider adminProvider, A
 
         try
         {
+            if (hasCustomScript)
+            {
+                logger.LogInformation("Executing custom creation script for database '{DatabaseName}':{NewLine}{Script}", databaseResource.DatabaseName, Environment.NewLine, script);
```
Comment (JamesNK) [inline src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs]: nit: This embeds NewLine as structured log information. I don't think it should be passed in like this.

File: src/Aspire.Hosting.MySql/MySqlBuilderExtensions.cs [hunk 2]
```diff
@@ -164,9 +164,11 @@ private static async Task CreateDatabaseAsync(MySqlConnection sqlConnection, MyS
             }
             else
             {
+                logger.LogInformation("Executing custom creation script for database '{DatabaseName}':{NewLine}{Script}", sqlDatabase.DatabaseName, Environment.NewLine, scriptAnnotation.Script);
```
Comment (JamesNK) [inline src/Aspire.Hosting.MySql/MySqlBuilderExtensions.cs]: nit: This embeds NewLine as structured log information. I don't think it should be passed in like this.

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs [hunk 3]
```diff
@@ -734,8 +734,21 @@ private static async Task CreateDatabaseAsync(NpgsqlConnection npgsqlConnection,
...
             var quotedDatabaseIdentifier = new NpgsqlCommandBuilder().QuoteIdentifier(npgsqlDatabase.DatabaseName);
             using var command = npgsqlConnection.CreateCommand();
-            command.CommandText = scriptAnnotation?.Script ?? $"CREATE DATABASE {quotedDatabaseIdentifier}";
-            await command.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
+            var commandText = scriptAnnotation?.Script ?? $"CREATE DATABASE {quotedDatabaseIdentifier}";
+            command.CommandText = commandText;
+
+            if (scriptAnnotation?.Script is { } script)
+            {
+                logger.LogInformation("Executing custom creation script for database '{DatabaseName}':{NewLine}{Script}", npgsqlDatabase.DatabaseName, Environment.NewLine, script);
```
Comment (JamesNK) [inline src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs]: nit: This embeds NewLine as structured log information. I don't think it should be passed in like this.

repeat for all

File: src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs (hunk 1 above)
Comment (JamesNK) [inline src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs]: Script is now being logged. Could it contain sensitive information that shouldn't be logged?

Applies to all places that script is logged in PR.

File: src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs [hunk 4]
```diff
@@ -312,10 +312,15 @@ private static async Task CreateDatabaseAsync(SqlConnection sqlConnection, SqlSe
...
-                    await command.ExecuteNonQueryAsync(ct).ConfigureAwait(false);
+                    var batch = batchBuilder.ToString();
+                    if (!string.IsNullOrWhiteSpace(batch))
+                    {
+                        batchNumber++;
+                        await ExecuteBatchAsync(batch, batchNumber, 1, ct).ConfigureAwait(false);
+                    }
                 }
+
+                logger.LogInformation("Completed custom creation script for database '{DatabaseName}'", sqlDatabase.DatabaseName);
```
Comment (JamesNK) [inline src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs]: Nit: The "Completed custom creation script" message fires unconditionally after the batch loop. If a user provides a script that is entirely empty/whitespace-only lines interspersed with GO statements, no batches actually execute (thanks to the `!string.IsNullOrWhiteSpace(batch)` guard), but the completion message still fires. Consider gating this on `batchNumber > 0`.

File: src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs [hunk 5]
```diff
@@ -324,5 +329,19 @@ private static async Task CreateDatabaseAsync(SqlConnection sqlConnection, SqlSe
...
+
+        async Task ExecuteBatchAsync(string batch, int batchNumber, int executionCount, CancellationToken cancellationToken)
+        {
+            logger.LogInformation("Executing custom creation script batch {BatchNumber} for database '{DatabaseName}':{NewLine}{Batch}", batchNumber, sqlDatabase.DatabaseName, Environment.NewLine, batch);
+
+            for (var i = 0; i < executionCount; i++)
+            {
+                using var command = sqlConnection.CreateCommand();
+                command.CommandText = batch;
+                var rowsAffected = await command.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
```
Comment (JamesNK) [inline src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs]: If a user passes `GO 1000` (execute batch N times), this produces N log messages — one per iteration. For high repeat counts this could flood the resource logger. Consider logging every Nth iteration (e.g., every 100th) or only the first, last, and a periodic sample, to keep output manageable while still providing visibility:

```csharp
for (var i = 0; i < executionCount; i++)
{
    using var command = sqlConnection.CreateCommand();
    command.CommandText = batch;
    var rowsAffected = await command.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);

    if (executionCount <= 10 || i == 0 || i == executionCount - 1 || (i + 1) % 100 == 0)
    {
        logger.LogInformation("Completed custom creation script batch {BatchNumber} execution {ExecutionNumber}/{ExecutionCount} for database '{DatabaseName}' with {RowsAffected} rows affected", batchNumber, i + 1, executionCount, sqlDatabase.DatabaseName, rowsAffected);
    }
}
```

**Review by JamesNK — COMMENTED** [review]
2 minor suggestions: (1) guard "Completed" log on at least one batch having executed, (2) consider throttling per-iteration log messages for high `GO N` repeat counts.

**Review by mitchdenny — COMMENTED** [review]
Addressed your feedback in the latest commit:

1. **NewLine as structured param / script content** — Removed the script/batch content and `{NewLine}` parameter from all 'Executing' log messages across all four providers (SQL Server, PostgreSQL, MySQL, Kusto). Script content could contain sensitive data and a newline is a formatting concern, not a structured data value.

2. **Sensitive data concern** — See above. We've removed the script body from all log messages to avoid the risk of logging credentials or other sensitive content.

3. **Completion log fires on empty script (SQL Server)** — Fixed. The 'Completed' message is now gated on `batchNumber > 0`, so it won't fire when a script consists entirely of whitespace between GO statements.

4. **GO N repeat count flooding the logger** — We'd like to keep per-iteration logging as-is. If a batch fails on iteration 500 of 1000, the per-iteration log gives the developer the precise execution number they need to diagnose the failure. Without it, they'd have no way to tell which iteration failed. We think the diagnostic value outweighs the verbosity concern for a developer-authored script path.

File: src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs]: Fixed — script content and `{NewLine}` have been removed from all four providers.

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs (hunk 3 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs]: Fixed — same as above.

File: src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs]: Good catch. We've removed the script body from all log messages across all four providers — the log now just says "Executing custom creation script for database 'X'" with no script content, so there's no risk of inadvertently logging credentials or other sensitive data.

File: src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs (hunk 4 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs]: Fixed — the "Completed" message is now gated on `batchNumber > 0`.

File: src/Aspire.Hosting.MySql/MySqlBuilderExtensions.cs (hunk 2 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.MySql/MySqlBuilderExtensions.cs]: Fixed — same as above, removed from all providers.

File: src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs (hunk 5 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs]: We'd prefer to keep the per-iteration logging as-is. If a batch fails mid-way through a high repeat count (e.g. iteration 500 of 1000), the per-iteration log gives the developer the exact execution number they need to diagnose the failure. Without it there's no way to tell which iteration failed — which defeats the purpose of the logging. We think that diagnostic value outweighs the verbosity concern.

**Comment by mitchdenny** [comment]
# PR Testing Report

## PR Information
- **PR Number:** #18138
- **Title:** Log custom creation script execution in resource logger for `WithCreationScript(...)`
- **Head Commit:** 855405593f8624981e7d38c0d3c9c3b4bceac109
- **Tested At:** 2026-06-12T03:59:00Z

## Artifact Version Verification
- **Expected Commit:** 855405593f8624981e7d38c0d3c9c3b4bceac109
- **Tested Source:** Local PR worktree at HEAD = 855405593 (rebuilt from source — no dogfood CLI install needed for a hosting-only change being tested in-repo via the playground)
- **Status:** ✅ Verified — local worktree HEAD matches PR head

## Changes Analyzed

### Files Changed
- `src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs`
- `src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs`
- `src/Aspire.Hosting.MySql/MySqlBuilderExtensions.cs`
- `src/Aspire.Hosting.Azure.Kusto/AzureKustoBuilderExtensions.cs`
- `tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs`
- `tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs`
- `tests/Aspire.Hosting.MySql.Tests/MySqlFunctionalTests.cs`

### Change Categories
- [ ] CLI changes
- [x] Hosting integration changes (SQL Server, PostgreSQL, MySQL, Azure Kusto)
- [ ] Dashboard changes (consumed-only — logs surface in resource console)
- [ ] CI infrastructure changes
- [ ] VS Code extension changes
- [x] Test changes (functional test assertions added)

## Test Scenarios Executed

### Scenario 1: MySQL `WithCreationScript` happy path (manual dashboard inspection)
**Objective:** Verify the new resource-logger feedback appears in the dashboard console for a database created with `WithCreationScript`, and that script content / `{NewLine}` are not emitted.
**Coverage Type:** Happy path + post-review-fix verification
**Status:** ✅ Passed

**Steps:**
1. Stopped previously running MySQL playground processes.
2. Rebuilt `playground/mysql/MySqlDb.AppHost` at HEAD = 855405593 (includes all post-review fixes).
3. Launched the playground AppHost (`dotnet run --no-build`).
4. Waited for MySQL container startup and database initialization.
5. Opened the Aspire dashboard at `https://localhost:59180` and navigated to the `mysql` resource console.

**Evidence (lines 60-67 of `mysql` resource console):**

```
60  Creating database 'catalog'
61  Database 'catalog' created successfully
62  Creating database 'myTestDb'
63  Database 'myTestDb' created successfully
64  Creating database 'myTestDb2'
65  Executing custom creation script for database 'myTestDb2'
66  Completed custom creation script for database 'myTestDb2' (2 rows affected)
67  Database 'myTestDb2' created successfully
```

`myTestDb2` is the database that has `.WithCreationScript(...)` in `playground/mysql/MySqlDb.AppHost/AppHost.cs`. The script ends with two `INSERT` statements, which is why "2 rows affected" appears.

**Observations:**

| Behaviour | Expected | Observed | Result |
|-----------|----------|----------|--------|
| New "Executing custom creation script" message appears for `WithCreationScript` databases | Yes | Yes (line 65) | ✅ |
| New "Completed custom creation script" message appears | Yes | Yes (line 66) | ✅ |
| Script content **not** logged (post-review fix for sensitive data concern) | Script body absent | No script content present | ✅ |
| `{NewLine}` token **not** embedded as a structured log parameter (post-review fix) | No raw `{NewLine}` text | No `{NewLine}` or stray newlines in message | ✅ |
| Rows-affected count shown when meaningful (post-review fix for `-1 rows affected`) | `(2 rows affected)` for INSERT statements | `(2 rows affected)` shown | ✅ |
| Default-path databases (no `WithCreationScript`) emit only the existing "Creating" / "created successfully" messages, **not** the new "Executing custom creation script" line | Silent for default path | `catalog` and `myTestDb` show only existing messages (lines 60-63), no new "Executing" line | ✅ |

This single scenario covers all four post-review fixes (sensitive data, `{NewLine}`, `-1 rows affected`) plus confirms the change is silent for the default code path.

## Coverage Not Exercised (by user request — "Quick" scope)
- SQL Server multi-batch `GO ...` script — covered by existing unit + functional tests; same logging pattern as MySQL.
- PostgreSQL `WithCreationScript` — same logging pattern, covered by existing functional test.
- Azure Kusto — requires Kusto emulator, not available locally; covered by unit tests.
- Failing-script scenario — pre-existing "Failed to create database" error path is unchanged by this PR.
- SQL Server `batchNumber > 0` gate (empty script edge case) — would require a contrived script of only whitespace between GO statements; behavior is straightforward in the diff.

## Summary
| Scenario | Status | Notes |
|----------|--------|-------|
| MySQL WithCreationScript dashboard inspection | ✅ Passed | All post-review fixes confirmed in live output |

## Overall Result
**✅ PR VERIFIED**

The latest commit (post-review fixes) produces clean, useful log output in the dashboard's resource console:
- No script content is logged (sensitive-data concern addressed).
- No raw `{NewLine}` tokens.
- Rows-affected count appears only when meaningful (DML INSERT/UPDATE, not DDL).
- The new logging is correctly gated on the `WithCreationScript` annotation — databases without a custom script are unaffected.

File: src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs [hunk 6]
```diff
@@ -734,8 +734,30 @@ private static async Task CreateDatabaseAsync(NpgsqlConnection npgsqlConnection,
         {
             var quotedDatabaseIdentifier = new NpgsqlCommandBuilder().QuoteIdentifier(npgsqlDatabase.DatabaseName);
             using var command = npgsqlConnection.CreateCommand();
-            command.CommandText = scriptAnnotation?.Script ?? $"CREATE DATABASE {quotedDatabaseIdentifier}";
-            await command.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
+            var commandText = scriptAnnotation?.Script ?? $"CREATE DATABASE {quotedDatabaseIdentifier}";
+            command.CommandText = commandText;
+
+            if (scriptAnnotation?.Script is { } script)
```
[the remark this answers is not in the document]
Comment (mitchdenny) [reply src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs]: Good catch — fixed in f2b1a3226 by switching to  to match the style of the existing check below.

**Review by JamesNK — APPROVED** [review]
